### PR TITLE
Highlight active menu item

### PR DESCRIPTION
**Pull request by ademuanthony**
Partial fix for #134
This PR worked on mempool, pow, vsp, exchange and propagation pages,
Make chart and table rendering to use a single page and brought the ability to persist all filters, both on table and chart view across url reuse.

image

### CODE REVIEW

**Comment by ademuanthony** [comment]
> image
> 
> Pr seems broken on exchange and vsp pages.

Exchange page is untouched.
Please build locally and test again

File: web/views/pow.html [hunk 1]
```diff
@@ -38,22 +38,28 @@
...
 
                            <div class="control-div pt-2" data-target="pow.chartSourceWrapper">
+                               {{ $selectedPools := .pow.selectedPools }}
                                 {{ range $index, $filter := .pow.powSource}}
                                 <div class="form-check form-check-inline" data-target="pow.poolDiv">
                                     <input data-target="pow.pool" data-action="click->pow#poolCheckChanged"
                                            class="form-check-input" type="checkbox" id="inlineCheckbox-{{$filter.Source}}"
-                                           value="{{$filter.Source}}" {{ if eq $index 0 }} checked {{ end }}>
+                                           value="{{$filter.Source}}" {{ if strListContains $selectedPools $filter.Source }} checked {{ end }}>
+
```
Comment (Sirmorrison) [inline web/views/pow.html]: Remove space.

File: web/handlers.go [hunk 2]
```diff
@@ -91,7 +116,7 @@ func (s *Server) fetchExchangeData(req *http.Request) (map[string]interface{}, e
 	}
 
 	filterInterval, err := strconv.Atoi(interval)
```
Comment (Sirmorrison) [inline web/handlers.go]: `filterInterval` will be better as `selectedInterval`.

File: web/handlers.go [hunk 3]
```diff
@@ -100,116 +125,96 @@ func (s *Server) fetchExchangeData(req *http.Request) (map[string]interface{}, e
...
 		"pageSizeSelector":     pageSizeSelector,
-		"currentPage":          pageToLoad,
-		"chartView":            false,
-		"selectedViewOption":   "table",
-		"previousPage":         pageToLoad - 1,
-		"totalPages":           int(math.Ceil(float64(totalCount) / float64(pageSize))),
-		"selectedExchange":     selectedExchange,
 		"selectedCurrencyPair": selectedCurrencyPair,
 		"selectedNum":          pageSize,
 		"selectedInterval":     filterInterval,
```
Comment (Sirmorrison) [inline web/handlers.go]: `filterInterval` should be `selectedInterval`

File: web/handlers.go [hunk 4]
```diff
@@ -100,116 +125,96 @@ func (s *Server) fetchExchangeData(req *http.Request) (map[string]interface{}, e
...
+	selectedInterval := req.FormValue("selected-interval")
+	selectedExchange := req.FormValue("selected-exchange")
 
 	data := map[string]interface{}{}
 
 	ctx := req.Context()
 	interval, err := strconv.Atoi(selectedInterval)
 	if err != nil {
-		s.renderErrorJSON(err.Error(), res)
+		s.renderErrorJSON(fmt.Sprintf("invalid intervale, %s", err.Error()), res)
```
Comment (Sirmorrison) [inline web/handlers.go]: `s.renderErrorJSON(fmt.Sprintf("invalid intervale, %s", err.Error()), res)` should be `s.renderErrorJSON(fmt.Sprintf("Invalid interval, %s", err.Error()), res)`

File: web/handlers.go [hunk 5]
```diff
@@ -981,17 +979,16 @@ func (s *Server) fetchBlockData(req *http.Request) (map[string]interface{}, erro
```
Comment (Sirmorrison) [inline web/handlers.go]: from line 971, change `strconv.ParseInt(page, 10, 32)` to `strconv.Atoi(page)` that will remove the conversion on line 976.

File: web/handlers.go [hunk 6]
```diff
@@ -1007,16 +1004,16 @@ func (s *Server) fetchBlockData(req *http.Request) (map[string]interface{}, erro
...
-		"totalPages":        int(math.Ceil(float64(totalCount) / float64(pageSize))),
+		"records":              voteSlice,
+		"currentPage":          pageToLoad,
+		"propagationRecordSet": propagationRecordSet,
+		"pageSizeSelector":     pageSizeSelector,
+		"selectedFilter":       "blocks",
+		"blocks":               true,
+		"selectedNum":          pageSize,
+		"url":                  "/blockdata",
+		"previousPage":         int(pageToLoad - 1),
```
Comment (Sirmorrison) [inline web/handlers.go]: the above fix will remove this conversion here and on line 1021.

File: web/handlers.go [hunk 7]
```diff
@@ -1104,16 +1101,16 @@ func (s *Server) fetchVoteData(req *http.Request) (map[string]interface{}, error
...
-		"totalPages":        int(math.Ceil(float64(totalCount) / float64(pageSize))),
+		"voteRecords":          voteSlice,
+		"currentPage":          pageToLoad,
+		"propagationRecordSet": propagationRecordSet,
+		"pageSizeSelector":     pageSizeSelector,
+		"selectedFilter":       "votes",
+		"votes":                true,
+		"selectedNum":          pageSize,
+		"url":                  "/votesdata",
+		"previousPage":         int(pageToLoad - 1),
```
Comment (Sirmorrison) [inline web/handlers.go]: Also the above will fix this also. and line 1118.

File: web/handlers.go [hunk 8]
```diff
@@ -1078,17 +1075,17 @@ func (s *Server) fetchVoteData(req *http.Request) (map[string]interface{}, error
```
Comment (Sirmorrison) [inline web/handlers.go]: line 1067 `strconv.ParseInt(page, 10, 32)` to `strconv.Atoi(page)`
the above will get rid of the conversion on line 1071

File: web/public/app/src/controllers/vsp_controller.js [hunk 9]
```diff
@@ -151,32 +174,29 @@ export default class extends Controller {
   }
 
   chartSourceCheckChanged (event) {
+    this.vsps = []
+    this.chartSourceTargets.forEach(chartSource => {
+      if (chartSource.checked) {
+        this.vsps.push(chartSource.value)
+      }
+    })
     this.fetchDataAndPlotGraph()
```
Comment (Sirmorrison) [inline web/public/app/src/controllers/vsp_controller.js]: line space.

File: web/public/app/src/controllers/vsp_controller.js [hunk 10]
```diff
@@ -151,32 +174,29 @@ export default class extends Controller {
...
-        _this.drawInitialGraph()
-        return
-      }
-
-      _this.plotGraph(result)
+    const _this = this
+    if (this.vsps.length === 0) {
+      return
+    }
+    const queryString = `data-type=${this.dataType}&vsps=${this.vsps.join('|')}&view-option=${_this.selectedViewOption}`
```
Comment (Sirmorrison) [inline web/public/app/src/controllers/vsp_controller.js]: line space.

File: web/public/app/src/controllers/vsp_controller.js [hunk 11]
```diff
@@ -210,12 +230,6 @@ export default class extends Controller {
         }
       }
     }
-    switch (this.graphTypeTarget.value) {
-      case 'Immature':
-
-        break
-    }
-
     _this.chartsView = new Dygraph(
```
Comment (Sirmorrison) [inline web/public/app/src/controllers/vsp_controller.js]: line space.

File: web/template.go [hunk 12]
```diff
@@ -48,5 +49,16 @@ func templateFuncMap() template.FuncMap {
...
 		},
+		"strListContains": func(stringList []string, needle string) bool {
+			for _, value := range stringList {
+				if value == needle {
+					return true
+				}
+			}
+			return false
+		},
+		"replace": func(input string, old string, new string) string {
```
Comment (Sirmorrison) [inline web/template.go]: `replace` ??? what are we replacing?

File: web/public/app/src/controllers/vsp_controller.js (hunk 9 above)
Comment (ademuanthony) [reply web/public/app/src/controllers/vsp_controller.js]: `no-trailing-space`

File: web/public/app/src/controllers/vsp_controller.js (hunk 11 above)
Comment (ademuanthony) [reply web/public/app/src/controllers/vsp_controller.js]: `no-trailing-pace`

File: postgres/exchange_ticks.go [hunk 13]
```diff
@@ -271,7 +271,7 @@ func (pg *PgDb) AllExchangeTicksInterval(ctx context.Context) ([]ticks.TickDtoIn
 func (pg *PgDb) ExchangeTicksChartData(ctx context.Context, selectedTick string, currencyPair string, selectedInterval int, source string) ([]ticks.TickChartData, error) {
 	exchange, err := models.Exchanges(models.ExchangeWhere.Name.EQ(source)).One(ctx, pg.db)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("the selected exchange, %s does not exist, %s", source, err.Error())
```
Comment (Sirmorrison) [inline postgres/exchange_ticks.go]: Please start sentences with capital letter. Repeat for all instances.

**Comment by Sirmorrison** [comment]
image

just notice when i select a combination of filter and there are no data to be displayed, the drop down options hide.

**Pull request closed without merging**